examples/classic_controllers/classic_controllers.py

```diff
@@ -632,113 +828,636 @@ def __init__(self, environment, stages, ref_states, color='g', **controller_kwar
+                                           vmin=np.nanmin(self.i_d_inter_plot), vmax=np.nanmax(self.i_d_inter_plot))
+                self.i_d_plot.set_ylabel(r'$\Psi / Vs$')
+                self.i_d_plot.set_xlabel(r'$T / Nm$')
+                self.i_d_plot.set_title(r'$i_\mathrm{d}(T, \Psi)$')
+
+    def solve_analytical(self, torque, psi):
+        """
+           Assuming linear magnetization characteristics, the optimal currents for given torque and flux can be obtained
+           by solving the torque and flux equations. These lead to a fourth degree polynomial which can be solved
+           analytically.  There are two ways to use this analytical solution for control. On the one hand, the currents
+           can be determined in advance as in the case of interpolation for different torques and fluxes and stored in a
+           LUT (torque_control='analytical'). On the other hand, the solution can be calculated at runtime with the
+           given torque and flux (torque_control='online').
+        """
+
+        poly = [self.l_d ** 2 * (self.l_d - self.l_q) ** 2,
+                2 * self.l_d ** 2 * (self.l_d - self.l_q) * self.psi_p + 2 * self.l_d * self.psi_p * (
+                        self.l_d - self.l_q) ** 2,
+                self.l_d ** 2 * self.psi_p ** 2 + 4 * self.l_d * self.psi_p ** 2 * (self.l_d - self.l_q) + (
+                        self.psi_p ** 2 - psi ** 2) * (
+                        self.l_d - self.l_q) ** 2,
+                2 * self.l_q * self.psi_p ** 3 + 2 * (self.psi_p ** 2 - psi ** 2) * self.psi_p * (self.l_d - self.l_q),
+                (self.psi_p ** 2 - psi ** 2) * self.psi_p ** 2 + (self.l_q * 2 * torque / (3 * self.p)) ** 2]
+
+        sol = np.roots(poly)
+        i_d = np.real(sol[-1])
+        i_q = 2 * torque / (3 * self.p * (self.psi_p + (self.l_d - self.l_q) * i_d))
+        return i_d, i_q
+
+    def get_i_d_q(self, torque, psi, psi_idx):
+        i_d, i_q = self.solve_analytical(torque, psi)
+        if i_d > self.mtpc[psi_idx, 1]:
+            i_d = self.mtpc[psi_idx, 1]
+            i_q = self.mtpc[psi_idx, 2]
+        return i_d, i_q
+
+    def get_t_idx(self, torque):
+        torque = np.clip(torque, self.t_min, self.t_max)
+        return int(round((torque - self.t_min) / (self.t_max - self.t_min) * (self.t_count - 1)))
+
+    def get_psi_idx(self, psi):
+        psi = np.clip(psi, self.psi_min, self.psi_max)
+        return int(round((psi - self.psi_min) / (self.psi_max - self.psi_min) * (self.psi_count - 1)))
+
+    def get_psi_idx_mtpf(self, psi):
+        return np.clip(int((self.psi_count - 1) - round(psi / self.psi_max_mtpf * (self.psi_count - 1))), 0,
+                       self.psi_count)
+
+    def get_t_idx_mtpc(self, torque):
+        return np.clip(int(round((torque + self.max_torque) / (2 * self.max_torque) * (self.t_count - 1))), 0,
+                       self.t_count)
+
+    def control(self, state, torque):
+
```

Comment:
For this central method, a docstring as well as some inline comments are necessary.